chat-widget/src/common/facades/FacadeChatSDK.ts

```diff
@@ -0,0 +1,284 @@
+    private token!: any | "";
+    private expiration = 0;
+    private isAuthenticated!: boolean;
+    private getAuthToken?: (authClientFunction?: string) => Promise<string | null>;
+
+    public getChatSDK(): OmnichannelChatSDK {
+        return this.chatSDK;
+    }
+
+    public destroy() {
+        this.token = null;
+        this.expiration = 0;
+    }
+
+    public isTokenSet() {
+        return !isNullOrEmptyString(this.token);
+    }
+
+    constructor(input: IFacadeChatSDKInput) {
+        this.chatSDK = input.chatSDK;
+        this.chatConfig = input.chatConfig;
+        this.getAuthToken = input.getAuthToken;
+        this.isAuthenticated = input.isAuthenticated;
+    }
+
+    private convertExpiration(expiration: number): number {
+
+        // token with no expiration, are not going to be validated
+        if (expiration === undefined){
+            return 0;
+        }
+        
+        const expStr = expiration?.toString();
+        if ((expStr.indexOf(".") > -1) && expStr.toString().length >= 13) {
+            return Math.floor(expiration / 1000);
+        }
+        return expiration;
+    }
+
+    private isTokenExpired(): boolean {
+
+        // if expiration is 0, token is not going to be validated ( this is to cover the case of token with no expiration)
+        if (this.expiration === 0){
+            return false;
+        }
+        
+        // obtain current time in seconds
+        const now = Math.floor(Date.now() / 1000);
+        // compare expiration time with current time
+        if (now > this.expiration) {
+            return true;
+        }
+        return false;
+    }
+
+    private async setToken(token: string | null): Promise<void> {
+        // token must be not null, and must be new
+        if (!isNullOrEmptyString(token) && token !== this.token) {
+            const instant = Math.floor(Date.now() / 1000);
+            this.token = token;
+            // decompose token
+            const tokenParts = this.token.split(".");
+            // decode token
+            const tokenDecoded = JSON.parse(atob(tokenParts[1]));
+            // calculate expiration time
+            this.expiration = this.convertExpiration(tokenDecoded.exp);
+            // this is a control , in case the getAuthToken function returns same token
+            if (this.expiration > 0 && (this.expiration < instant)) {
+
+                TelemetryHelper.logFacadeChatSDKEvent(LogLevel.ERROR, {
+                    Event: TelemetryEvent.NewTokenExpired,
+                    Description: "New token is already expired",
+                    ExceptionDetails: {
+                        "Instant": instant,
+                        "Expiration": this.expiration
+                    }
+                });
+                throw new Error("New token is already expired, with epoch time " + this.expiration);
+            }
+        }
+    }
+
+    private async tokenRing(): Promise<PingResponse> {
+        if (this.isAuthenticated) {
```

Comment:
Using early returns can help avoid excessive nesting

```typescript
if (!this.isAuthenticated) return { result: true, message: "Authentication no needed" };
```

Same with the next if condition `if (!this.isTokenSet() || this.isTokenExpired() ) {`